### ISSUE TITLE
Have you in plan change gui from gtk to other like juice or something cross-platform?

### ISSUE THREAD

**Issue by proboofficial** [calf-studio-gear/calf#346]
New version of Ardour 8.4 dont want work with calf plugins. Problem is too old gtk version

**Comment by darinrogers** [comment]
For those searching, I _believe_ the error you'll see is "[ERROR]: failed to instantiate LV2 GUI".

**Comment by proboofficial** [comment]
Yes is this error, but I think good idea is make calf studio gear cross platform tool

**Comment by bratpeki** [comment]
Either:

- **using a new GUI framework** (Qt, GTK+, GL/Vuklan,...),
- **reverting to a generic LADSPA interface** or
- **using a new plugin framework** altogether (stuff like DPF and JUCE).

**Comment by bratpeki** [comment]
In the meantime, the Debian Multimedia Maintainers are hosting [calf-ladspa](https://packages.debian.org/buster/sound/calf-ladspa), which is GUI-less, but fully functional.

**Comment by trebmuh** [comment]
See also: https://github.com/calf-studio-gear/CTK

**Comment by boomshop** [comment]
Those problems were already in sight a couple of years ago, that's why I started CTK - as a replacement for GTK2. Basic stuff already works afair, things like composing, layering, event bubbling, clipped region redraw and the like. Still a long way to go. After CTK is done, Calf needs to adapt to the new toolkit. Since it has grown into this huge pile of "devices" and still suffers from a few design flaws UI-wise (e.g. direct access between UI and DSP needed) it's nothing to be done on a Sunday evening but more like an effort of several man months. I unfortunately don't have time - neither for Calf nor for CTK - anymore, so I fear this is the nail in the coffin as of yet.

Bummer, since it's still the only plug-ins I'm using in my studio.

**Comment by falkTX** [comment]
I have a proposal that can buy us a few more years time - starting the gtk UI out of process and doing X11 window reparenting.
The idea being that the lv2 ui object would be a fake one that starts an external process with the real gtk UI, and use IPC to send any received data to it and vice-versa.
We can use LV2 X11 UI type to show the UI as embedded, giving a similar feeling to before, but without conflicts of e.g. host using gtk3 or a modified gtk2 like ardour.

This would allow to load the calf UI, assuming the linux distro still ships with gtk2 of course.
It is hard to do and gives quick results

The main problem here is calf plugin graphs being tied to their DSP side, if we start the UI in an external process we lose access to them.
If that is an acceptable trade-off, I can come up with something quite quick to allow to load the current UIs

**Comment by falkTX** [comment]
@boomshop I was going to reply to your email, but I was too curious to see if my ideas could work well enough so I ended up creating https://github.com/falkTX/lv2-gtk-ui-bridge in the last 2 days :tada: 

This allows us to use lv2 gtk2 uis on any host that supports x11 ui type.
Verified to work on latest Ardour with Calf, EQ10Q stuff and Invada.
Also works for loading gtk2 uis in jalv.gtk3 :sweat_smile: 
Does not work in Reaper though, because it does not seem to support "out of main bundle UI" kinda deal that I setup on that repo. But that is mostly a technicality, if we define the UI inside the calf.lv2 bundle it will load the UI

![image](https://github.com/calf-studio-gear/calf/assets/1334853/c509ddfb-989e-4a2b-9e3e-d39f728118a0)

Initially I thought about something that would be useful just for Calf project, but ended up with a generic approach, so that is why it also works for EQ10Q and Invada.
Can easily add support for more, this uses a hardcoded list of plugin URIs to give the extra X11 UI to.

Anyway, testing and feedback is appreciated and all that.
Quite minimal deps, so should be easy enough to build.

PS: A side effect from this is that we can reuse it for more esoteric usecases too, say gtk3-webkit to have a web view as plugin UI. the process separation means we can use whatever libs we want without worry, and that includes any gtk or web view version

**Comment by boomshop** [comment]
@falkTX just wow. That's genius! Will test it and report back any problems once I find the time.

> if we define the UI inside the calf.lv2 bundle it will load the UI

Are there any implications to this in your opinion? Other than having lv2-gtk-ui-bridge as a dependency. E.g. breaking the homebrew version for MacOS, not having the bridge packaged for the various OS yet, something along those lines. We could also offer a set of manifests for Reaper users, or a script manipulating them.

**Comment by falkTX** [comment]
> > if we define the UI inside the calf.lv2 bundle it will load the UI
> 
> Are there any implications to this in your opinion? Other than having lv2-gtk-ui-bridge as a dependency. E.g. breaking the homebrew version for MacOS, not having the bridge packaged for the various OS yet, something along those lines. We could also offer a set of manifests for Reaper users, or a script manipulating them.

I think the best course of action is to eventually have this special bundle be part of calf, I think that makes the most sense on the "long"-term.
I still need to finalize the windows part of the IPC stack, it is something I had a few issues before. macOS and Linux/POSIX already works though.
So as I see it, we let this be a separate project for just a bit to do quick experiments with, and once stable we can move things into the calf project

**Comment by falkTX** [comment]
@boomshop were you able to try it? I finalized the windows IPC parts though for this gtk2-external-in-x11 thing is not that useful since it relies on x11 embedding.

if we would be integrating it into calf I need some help as my knowledge of autoconf/autotools is very lacking.
very roughly the steps would be:
1. add the project source files in this repo
2. set autotools to build those files if detecting x11 linux/bsd system (and lv2 gui is on)
3. set the build so that it creates 2 new objects: 1 shared library for the x11 gui, one executable for the gui wrapper
4. tweak the lv2 ttl generator to add the x11 gui details
5. tweak the copied ipc code to minimize to calf needs (the C files, for no gtk3 or lilv needed)

I cannot do the step 2 and 3, but can do the others

**Comment by boomshop** [comment]
@falkTX sorry, I'm totally drowning in work atm (we have a huge contest for smoke divers upcoming where I'm the only developer voluntarily building the software to handle the whole contest) so I didn't have time for literally _anything_ unfortunately. Let's see... @kfoltman are you still around - maybe able to give a helping hand? Would be ace!

**Comment by falkTX** [comment]
@boomshop no worries, I might go ahead and do a first beta release for the lv2 bridge project then.

**Comment by x42** [comment]
Keep in mind that calf is still on the list of packages to be removed with the next debian release (when gtk2 is dropped):

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=947713

**Comment by falkTX** [comment]
> Keep in mind that calf is still on the list of packages to be removed with the next debian release (when gtk2 is dropped):
> 
> https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=947713

sounds like a good reason to release it now before it becomes completely worthless.

is gtk3 going the same path too?

**Comment by JohannesLorenz** [comment]
@falkTX I currently planned a CALF bugfix release [0.90.4](https://github.com/calf-studio-gear/calf/milestone/1) in the next months. Should the gtk change be included or separate?

Also, FYI, on branch "github-actions", I set up CMake for CALF which already works. If that helps you with anything, feel free to use my CMakeLists instead of automake.

**Comment by falkTX** [comment]
> @falkTX I currently planned a CALF bugfix release [0.90.4](https://github.com/calf-studio-gear/calf/milestone/1) in the next months. Should the gtk change be included or separate?

feels too much of a change to include in one go, specially since a release has not been made in a long time.

the cmake side for sure helps to setup the build to have this extra x11 embed option. are you wishing to proceed with cmake for real or just testing the waters here?

**Comment by x42** [comment]
> is gtk3 going the same path too?

Yeah, you'll have a few more years. Then again https://github.com/lv2/lv2/wiki#avoid-gui-frameworks

If this project is to continue, calf should first get its act together and fix the countless DSP issues. That's my 2c.

**Comment by suedwestlicht** [comment]
Well, as a non-programmer and user of Calf plugins my biggest worry are not some DSP bugs, but the impending inability to use them due to UI-toolkit issues.

They are easy to use, beautiful and sound good - even if, for example, the filters of the multiband plug-ins change the sound or the transient designer pops after signal pauses (a very quiet signal immediately before the "real" signal provides a remedy here).

**Comment by JohannesLorenz** [comment]
> > @falkTX I currently planned a CALF bugfix release [0.90.4](https://github.com/calf-studio-gear/calf/milestone/1) in the next months. Should the gtk change be included or separate?
> 
> feels too much of a change to include in one go, specially since a release has not been made in a long time.

OK, in this time I would just ask you to park your changes in a PR before moving to master, until we have 0.90.4 released.

> the cmake side for sure helps to setup the build to have this extra x11 embed option. are you wishing to proceed with cmake for real or just testing the waters here?

For real. On the `github-actions` branch, both are already equally good. I think CMake should soon be the prior build system, and, as for me, automake should be removed in the near future (to avoid making all changes in 2 build systems).

**Comment by your-friend-alice** [comment]
an additional perspective: the signal processing issues make sense to me, that seems like a good reason to move away from CALF. However, I don't have a time machine, and I already have many project files that do use CALF, often for synths and as part of non-trivial filter chains that can't be perfectly reproduced. I start things, run out of steam, sit on them for sometimes years, and come back to them, so backwards compatibility is pretty important to me!

I'm not a C++ developer and am not very familiar with the details of linking and stuff, is there not a way to just statically link the GTK2 dependency and prevent exporting any of GTK's symbols to avoid clashes? I wouldn't mind the wasted memory in exchange for not losing many hours of work

**Comment by tl072** [comment]
> I don't have a time machine, and I already have many project files that do use CALF, often for synths and as part of non-trivial filter chains that can't be perfectly reproduced. I start things, run out of steam, sit on them for sometimes years, and come back to them, so backwards compatibility is pretty important to me!

This is really the core of the issue I think. I am in the same boat. I cannot insist that people do coding for something that I am not sure I am even capable of doing (without a lot of new learning for a becoming obsolete framework), but hopefully my experience and @your-friend-alice are inspiring enough to keep things moving :) The DSP questions probably have some merit given the knowledge of the people I have seen making reference to them, but for what it is worth, I have really enjoyed using the calf plugins and i think they sound good for whatever it is that do :)

**Comment by ArnoJordan1** [comment]
I´d love to try calf plugins, but its not showing any UI in Reaper and other dawś because the gtk2 stuff. Is there a plan to change this ?

**Comment by suedwestlicht** [comment]
CALF plugins work fine in Qtractor.

If you need them in Reaper then try:
- Make sure that libgtk2 is installed on your system
- Install Carla and try to load CALF Plugins
- If that works then load Carla as a plugin into Reaper and load CALF plugins into Carla

**Comment by falkTX** [comment]
alternatively use https://github.com/falkTX/lv2-gtk-ui-bridge
but that doesnt help in reaper as it does not support split dsp/ui bundles

**Comment by ArnoJordan1** [comment]
unfortunately i am using yabridge, not carla, and i think can not run them together at the same time. But anyway, after upgrading wine the UI of all my plugins are frozen.
I really hope one day this will change. Linux is  much nicer to work with, but as long as the plugin industry does not support native linux plugins  we will not be able to use Linux for real pro audio work and have to stick to operating systems like windows 11 that become more and more unusable.

**Comment by juzzlin** [comment]
I have been using Calf (LV2) with Audacity on Ubuntu 24.04 LTS. I have disabled the UI anyway, because it doesn't work with the Wayland session. I'm still sad to read this. Everything is a mess nowadays.

**Comment by arnowaschk** [comment]
Did you consider DPF or JUCE?

**Comment by boomshop** [comment]
Sneak-peek of fresh project:

<img width="1425" height="954" alt="Image" src="https://github.com/user-attachments/assets/b6d2f8ad-cfca-4783-bd83-21cb7932ae6a" />

**Comment by suedwestlicht** [comment]
Beautiful as everything by calf. I'd love to see a complete calfNXT suite. I didn't find a replacement for calf Deesser, calf Transient Designer and calf Exciter.

A dynamic function in the eq would be great as well.

However thanks for letting us know.

**Comment by boomshop** [comment]
> Beautiful as everything by calf.

Thanks!

> I'd love to see a complete calfNXT suite.

That's what I'm working on atm. New concept, new UI, overhauled DSP, different focus.

> I didn't find a replacement for calf Deesser, calf Transient Designer and calf Exciter.

I'm also missing Limiter, MultibandLimiter, BassEnhancer, Reverb, VintageDelay and some others, that's why I have to take some action. Calf has always defined my personal sound.

> A dynamic function in the eq would be great as well.

Planned for.

> However thanks for letting us know.

You're welcome, although please don't expect anything useful next week or month.

**Comment by suedwestlicht** [comment]
As Debian Trixie still includes GTK2, there’s no need to rush.

**Comment by boomshop** [comment]
I'm on CachyOS now, so no chance unfortunately. And I want to start a new production, mix/master will take place this autumn/winter, so it's time for something new.

**Comment by boomshop** [comment]
@suedwestlicht  Lately I stumbled over this gem: #183 - had a good chuckle :)

If you promise to not flood the issues, maybe try these guys: https://github.com/boomshop/calfnxt
There already is Equalizer (w/ dynamic feature and various fixes in the dsp), Stereo (former Stereo Tools, now w/ correlation and gonio meter and decorrelation in the side part of the m/s matrix for smooth stereo enhancement) and Transients (former Transient Designer w/ various fixes in the dsp). More to come.

**Comment by suedwestlicht** [comment]
I built it on debian trixie.
- "cd ui && npm run dev" lets me play with the gui but with no connection to a running plugin. Parms are not changed there, I guess that's how it's designed
- I can load the eq into non-mixer-xt and carla. I can edit the parms with the generic controls and it works.
- But opening the plugin's gui draws a window without contents. 
  Unfortunately I didn't get any error messages.

**Comment by boomshop** [comment]
@suedwestlicht could you please put this in the calfnxt repo?

**Comment by x42** [comment]
...and here I was thinking calf would mature int cow or bull :)
